gax/retry_policy.h

```diff
@@ -48,21 +48,37 @@ class RetryPolicy {
    * @return true if the RPC operation should be retried.
    */
   virtual bool OnFailure(Status const& status) = 0;
+
+  /**
+   * Calculate the deadline for the next RPC operation.
+   *
+   * Any internal state modification, if neccessary, should occur in OnFailure.
+   *
+   * Note: this is different from the deadline in LimitedDurationRetryPolicy,
+   *       which is the deadline after which retry attempts should be abandoned.
+   *
+   * @return the _deadline_ for the next RPC, NOT its maximum _duration_.
+   */
+  virtual std::chrono::system_clock::time_point OperationDeadline() const = 0;
```

Comment:
As I recall the design was to have the policies setup the context?  Shouldn't this be something like:

```C++
  virual void Setup(gax::Context& context) const = 0;
```

?

Reply:
The problem I ran into there is that CallContext now stores pointers to RetryPolicy objects, so we run into the problem of circular includes.
I can do some forward declarations of either RetryPolicy or CallContext, move implementation code around into cc files, and let the linker take care of the problem, but that contradicts the style guide and feels wrong.

Reply:
Stepping back a little, is there a design doc that details how the initial interfaces for `g::c::bigtable::RPCRetryPolicy` and `g::c::bigtable::RPCBackoffPolicy` came about? I'd like to push back a little on the `Setup` method in favor of something more value-oriented, but I'm aware that that breaks with the existing semantics of google-cloud-cpp.

Reply:
Per out IM discussion, we can resolve this in a separate design doc and PR.